bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/JobManager.java

```diff
@@ -1100,19 +1101,25 @@ boolean join(InternalJobGroup jobGroup, long timeout, IProgressMonitor monitor)
 	}
 
 	/**
-	 * Returns a non-null progress monitor instance.  If the monitor is null,
-	 * returns the default monitor supplied by the progress provider, or a
-	 * NullProgressMonitor if no default monitor is available.
+	 * Returns a non-null progress monitor instance. If a progress provider is set,
+	 * it is asked to provide a wrapped instance of the given <code>monitor</code>
+	 * parameter. If no progress provider is set and <code>monitor</code> is null, a
+	 * NullProgressMonitor is returned in all other cases the given
+	 * <code>monitor</code> instance.
+	 *
+	 * @param monitor the monitor instance (could be null) for further usage
+	 * @return a non-null progress monitor instance.
 	 */
 	private IProgressMonitor monitorFor(IProgressMonitor monitor) {
-		if (monitor == null || (monitor instanceof NullProgressMonitor)) {
-			if (progressProvider != null) {
-				try {
-					monitor = progressProvider.getDefaultMonitor();
-				} catch (Exception e) {
-					String msg = NLS.bind(JobMessages.meta_pluginProblems, JobManager.PI_JOBS);
-					RuntimeLog.log(new Status(IStatus.ERROR, JobManager.PI_JOBS, JobManager.PLUGIN_ERROR, msg, e));
+		if (progressProvider != null) {
+			try {
+				IProgressMonitor monitorFor = progressProvider.monitorFor(monitor);
+				if (monitorFor == null) {
+					throw new IllegalStateException("ProgressProvider " + progressProvider + " returned null!"); //$NON-NLS-1$ //$NON-NLS-2$
```

Comment:
Here also please add the original monitor to the error message
"ProgressProvider " + progressProvider + " returned null for monitor: " + monitor